include/vcpkg/metrics.h

```diff
@@ -90,32 +97,92 @@ namespace vcpkg
 
     extern const std::array<BoolMetricEntry, static_cast<size_t>(BoolMetric::COUNT)> all_bool_metrics;
 
-    struct Metrics
+    // Batches metrics changes so they can be submitted under a single lock acquisition or
+    // in a single JSON payload.
+    struct MetricsSubmission
     {
-        Metrics() = default;
-        Metrics(const Metrics&) = delete;
-        Metrics& operator=(const Metrics&) = delete;
+        void track_elapsed_us(double value);
+        void track_buildtime(StringView name, double value);
+        void track_define(DefineMetric metric);
+        void track_string(StringMetric metric, StringView value);
+        void track_bool(BoolMetric metric, bool value);
+        void merge(MetricsSubmission&& other);
+
+        double elapsed_us = 0.0;
+        std::map<std::string, double, std::less<>> buildtimes;
+        std::set<DefineMetric> defines;
+        std::map<StringMetric, std::string> strings;
+        std::map<BoolMetric, bool> bools;
+    };
+
+    // Collects metrics, potentially from multiple threads.
+    // Member functions of this type are safe to call from multiple threads, and will
+    // be observed in a total order.
+    struct MetricsCollector
+    {
+        MetricsCollector() = default;
+        MetricsCollector(const MetricsCollector&) = delete;
+        MetricsCollector& operator=(const MetricsCollector&) = delete;
+        ~MetricsCollector() = default;
+
+        // Track
+        void track_elapsed_us(double value);
+        void track_buildtime(StringView name, double value);
+        void track_define(DefineMetric metric);
+        void track_string(StringMetric metric, StringView value);
+        void track_bool(BoolMetric metric, bool value);
+        void track_submission(MetricsSubmission&& submission_);
```

Comment:
This one intentionally had a different name because I don't think "merge" carries the right connotation; it isn't 2 things of the same kind that are being merged together, it's generically a batch metrics submission, and that it happens to be implemented in terms of merge under the covers right now is an implementation detail.

There is intentionally no API reason why MetricsCollector and MetricsSubmission need to be the same data structure.